### PR TITLE
Defect/dnn-17875

### PR DESCRIPTION
**Pull request by mean2me**

### CODE REVIEW

File: src/Modules/Manage/Dnn.PersonaBar.Users/Components/UsersController.cs [hunk 1]
```diff
@@ -488,6 +488,20 @@ private static bool IsAdmin(UserInfo user, PortalSettings portalSettings)
             return user.IsSuperUser || user.IsInRole(portalSettings.AdministratorRoleName);
         }
 
+        private static string applyWildcard(string searchText)
```
Comment (tpluscode) [inline src/Modules/Manage/Dnn.PersonaBar.Users/Components/UsersController.cs]: Please keep in mind that C# uses `PascalCase`

File: src/Modules/Manage/Dnn.PersonaBar.Users/Components/Prompt/Commands/ListUsers.cs [hunk 2]
```diff
@@ -96,13 +96,13 @@ public override ConsoleResultModel Run()
...
                 // do username lookup
-                var searchTerm = Username.Replace("%", "").Replace("*", "");
+                var searchTerm = Username.Trim().TrimEnd(new char[] { '%', '*' }); // Username.Replace("%", "").Replace("*", "");
                 getUsersContract.SearchText = searchTerm;
             }
             else if (!string.IsNullOrEmpty(Email))
             {
                 // must be email
-                var searchTerm = Email.Replace("%", "").Replace("*", "");
+                var searchTerm = Email.Trim().TrimEnd(new char[] { '%', '*' }); //.Replace("%", "").Replace("*", "");
```
Comment (tpluscode) [inline src/Modules/Manage/Dnn.PersonaBar.Users/Components/Prompt/Commands/ListUsers.cs]: Ditto

File: src/Modules/Manage/Dnn.PersonaBar.Users/Components/Prompt/Commands/ListUsers.cs [hunk 3]
```diff
@@ -96,13 +96,13 @@ public override ConsoleResultModel Run()
             if (!string.IsNullOrEmpty(Username))
             {
                 // do username lookup
-                var searchTerm = Username.Replace("%", "").Replace("*", "");
+                var searchTerm = Username.Trim().TrimEnd(new char[] { '%', '*' }); // Username.Replace("%", "").Replace("*", "");
```
Comment (tpluscode) [inline src/Modules/Manage/Dnn.PersonaBar.Users/Components/Prompt/Commands/ListUsers.cs]: I think this should be changed so that leading/trailing asterisk is changed to the percent sign.

**Review by tpluscode — CHANGES_REQUESTED** [review]
Please see my comment in Jira before proceeding

File: src/Modules/Manage/Dnn.PersonaBar.Users.Tests/UserControllerSearchTest.cs [hunk 4]
```diff
@@ -0,0 +1,82 @@
...
+
+            Assert.AreEqual(usersCtrl.lastSearch, expectedFilteredText);
+
+        }
+
+        private class UsersControllerTestable : UsersController
+        {
+            private Mock<IDataReader> dataReader;
+            
+            public string lastSearch { get; set; }
```
Comment (tpluscode) [inline src/Modules/Manage/Dnn.PersonaBar.Users.Tests/UserControllerSearchTest.cs]: Please also keep properties in PascalCase

File: src/Modules/Manage/Dnn.PersonaBar.Users/Components/Helpers/SearchTextFilter.cs [hunk 5]
```diff
@@ -0,0 +1,111 @@
+﻿using System.Text;
+using System.Text.RegularExpressions;
+
+namespace Dnn.PersonaBar.Users.Components.Helpers
+{
+    public class SearchTextFilter
+    {
+        
+        private static string getInStringSearchPattern(string searchText)
```
Comment (tpluscode) [inline src/Modules/Manage/Dnn.PersonaBar.Users/Components/Helpers/SearchTextFilter.cs]: Please make methods PascalCase

File: src/Modules/Manage/Dnn.PersonaBar.Users/Components/Helpers/SearchTextFilter.cs [hunk 6]
```diff
@@ -0,0 +1,111 @@
...
+                if (matchText != null && !string.IsNullOrEmpty(matchText))
+                {
+                    pattern.Append(matchText.Replace("*", "").Replace("%", ""));
+                    pattern.Append("%");
+                }
+            }
+            return pattern.ToString();
+        }
+
+        public static string Apply(string searchText)
```
Comment (tpluscode) [inline src/Modules/Manage/Dnn.PersonaBar.Users/Components/Helpers/SearchTextFilter.cs]: Please keep public methods on top

File: src/Modules/Manage/Dnn.PersonaBar.Users.Tests/UserControllerSearchTest.cs [hunk 7]
```diff
@@ -0,0 +1,82 @@
...
+
+            override protected IDataReader CallGetUsersBySearchTerm(
+                GetUsersContract usersContract,
+                bool? includeAuthorized,
+                bool? includeDeleted,
+                bool? includeSuperUsers)
+            {
+
```
Comment (tpluscode) [inline src/Modules/Manage/Dnn.PersonaBar.Users.Tests/UserControllerSearchTest.cs]: While we're being thorough, please avoid unnecessary whitespace

File: src/Modules/Manage/Dnn.PersonaBar.Users.Tests/UserControllerSearchTest.cs [hunk 8]
```diff
@@ -0,0 +1,82 @@
...
+        }
+
+        private class UsersControllerTestable : UsersController
+        {
+            private Mock<IDataReader> dataReader;
+            
+            public string lastSearch { get; set; }
+
+            override protected IDataReader CallGetUsersBySearchTerm(
```
Comment (tpluscode) [inline src/Modules/Manage/Dnn.PersonaBar.Users.Tests/UserControllerSearchTest.cs]: Typically it's `protected override`

File: src/Modules/Manage/Dnn.PersonaBar.Users.Tests/UserControllerSearchTest.cs [hunk 9]
```diff
@@ -0,0 +1,82 @@
...
+                GetUsersContract usersContract,
+                bool? includeAuthorized,
+                bool? includeDeleted,
+                bool? includeSuperUsers)
+            {
+
+                lastSearch = SearchTextFilter.Apply(usersContract.SearchText);
```
Comment (tpluscode) [inline src/Modules/Manage/Dnn.PersonaBar.Users.Tests/UserControllerSearchTest.cs]: This is not perfect, because the tests won't really verify what the "real" `UsersController` does.

File: src/Modules/Manage/Dnn.PersonaBar.Users/Components/UsersController.cs [hunk 10]
```diff
@@ -488,24 +490,39 @@ private static bool IsAdmin(UserInfo user, PortalSettings portalSettings)
...
+                    includeSuperUsers));
+
+            totalRecords = records.Count == 0 ? 0 : records[0].TotalCount;
+            return records;
+        }
+
+        protected virtual IDataReader CallGetUsersBySearchTerm(GetUsersContract usersContract,
+            bool? includeAuthorized, bool? includeDeleted, bool? includeSuperUsers)
+        {
+            var parsedSearchText = string.IsNullOrEmpty(usersContract.SearchText) ? "" : SearchTextFilter.Apply(usersContract.SearchText.Trim());
```
Comment (tpluscode) [inline src/Modules/Manage/Dnn.PersonaBar.Users/Components/UsersController.cs]: To follow up on the other comment, I was expecting that the filtered text would become a parameter of `CallGetUsersBySearchTerm`. This way you only wrap the `DataProvider.Instance().ExecuteReader()` call and nothing else.

```c#
var parsedSearchText = SearchTextFilter.Apply(usersContract.SearchText.Trim());

CallGetUsersBySearchTerm(
    usersContract,
    parsedSearchText
    includeAuthorized, 
    includeDeleted, 
    includeSuperUsers));
```

This way you will be able to verify that the controller passes the expected search text to the stored procedure.

File: src/Modules/Manage/Dnn.PersonaBar.Users/Components/Helpers/SearchTextFilter.cs (hunk 6 above)
Comment (tpluscode) [reply src/Modules/Manage/Dnn.PersonaBar.Users/Components/Helpers/SearchTextFilter.cs]: And maybe rename to something like `CleanWildcards`?

**Review by tpluscode — CHANGES_REQUESTED** [review]

File: src/Modules/Manage/Dnn.PersonaBar.Users.Tests/Properties/AssemblyInfo.cs [hunk 11]
```diff
@@ -0,0 +1,20 @@
...
+[assembly: AssemblyTrademark("")]
+[assembly: AssemblyCulture("")]
+
+[assembly: ComVisible(false)]
+
+[assembly: Guid("e6aefc23-3777-452c-8958-2c17d62584b5")]
+
+// [assembly: AssemblyVersion("1.0.*")]
+[assembly: AssemblyVersion("1.0.0.0")]
+[assembly: AssemblyFileVersion("1.0.0.0")]
```
Comment (tpluscode) [inline src/Modules/Manage/Dnn.PersonaBar.Users.Tests/Properties/AssemblyInfo.cs]: Please leave only the minimum attributes, as in the other `AssemblyInfo.cs` files and link `SolutionInfo.cs`

**Review by tpluscode — CHANGES_REQUESTED** [review]

File: src/Modules/Manage/Dnn.PersonaBar.Users.Tests/UserControllerSearchTest.cs (hunk 9 above)
Comment (mean2me) [reply src/Modules/Manage/Dnn.PersonaBar.Users.Tests/UserControllerSearchTest.cs]: Let's find another name.

File: src/Modules/Manage/Dnn.PersonaBar.Users/Components/UsersController.cs (hunk 10 above)
Comment (mean2me) [reply src/Modules/Manage/Dnn.PersonaBar.Users/Components/UsersController.cs]: Filtered test is an attribute of UsersContract. Don't need to make it an argument, IMO.

Comment (mean2me) [reply src/Modules/Manage/Dnn.PersonaBar.Users/Components/UsersController.cs]: Inside UsersController.CallGetUsersBySearchTerm() we actually call the DataProvider. Nothing else happens in the middle, so I think it's enough to test the value of *parsedSearchText* that way.
We care of the resulting string actually.
``` c#
protected virtual IDataReader CallGetUsersBySearchTerm(GetUsersContract usersContract,
            bool? includeAuthorized, bool? includeDeleted, bool? includeSuperUsers)
        {
            var parsedSearchText = string.IsNullOrEmpty(usersContract.SearchText) ? "" : SearchTextFilter.CleanWildcards(usersContract.SearchText.Trim());

            return DataProvider.Instance().ExecuteReader(
                    "Personabar_GetUsersBySearchTerm",
                    usersContract.PortalId,
                    string.IsNullOrEmpty(usersContract.SortColumn) ? "Joined" : usersContract.SortColumn,
                    usersContract.SortAscending,
                    usersContract.PageIndex,
                    usersContract.PageSize,
                    parsedSearchText,
                    includeAuthorized,
                    includeDeleted,
                    includeSuperUsers);
        }
```

File: src/Modules/Manage/Dnn.PersonaBar.Users.Tests/Properties/AssemblyInfo.cs [hunk 12]
```diff
@@ -0,0 +1,6 @@
+using System.Reflection;
+using System.Runtime.InteropServices;
+
+[assembly: AssemblyTitle("Dnn.PersonaBar.Users.Tests")]
+[assembly: ComVisible(false)]
+[assembly: Guid("e6aefc23-3777-452c-8958-2c17d62584b5")]
```
Comment (tpluscode) [inline src/Modules/Manage/Dnn.PersonaBar.Users.Tests/Properties/AssemblyInfo.cs:6]: Please remember to link the `SolutionInfo.cs`

File: src/Modules/Manage/Dnn.PersonaBar.Users.Tests/Dnn.PersonaBar.Users.Tests.csproj [hunk 13]
```diff
@@ -0,0 +1,94 @@
...
+    <Reference Include="System.Data" />
+    <Reference Include="System.Threading.Tasks.Extensions, Version=4.1.0.0, Culture=neutral, PublicKeyToken=cc7b13ffcd2ddd51, processorArchitecture=MSIL">
+      <HintPath>..\..\..\packages\System.Threading.Tasks.Extensions.4.3.0\lib\portable-net45+win8+wp8+wpa81\System.Threading.Tasks.Extensions.dll</HintPath>
+    </Reference>
+    <Reference Include="System.ValueTuple, Version=4.0.2.0, Culture=neutral, PublicKeyToken=cc7b13ffcd2ddd51, processorArchitecture=MSIL">
+      <HintPath>..\..\..\packages\System.ValueTuple.4.4.0\lib\netstandard1.0\System.ValueTuple.dll</HintPath>
+    </Reference>
+  </ItemGroup>
+  <ItemGroup>
+    <Compile Include="SearchUsersBySearchTermTest.cs" />
```
Comment (tpluscode) [inline src/Modules/Manage/Dnn.PersonaBar.Users.Tests/Dnn.PersonaBar.Users.Tests.csproj]: Did you forget to commit this file?

**Review by tpluscode — CHANGES_REQUESTED** [review]

**Review by tpluscode — APPROVED** [review]

**Comment by tpluscode** [comment]
fixes #565
